File: decision.cpp
#include "decision.h"

Decision::Decision()
{
    changePermission = false;
    firstTime = true;
}

Decision::~Decision()
{

}

vector<robot> Decision::getTeamRobots()
{
    return teamRobot;
}

Point2f Decision::getCentroidDef()
{
    return centroidDef;
}

Point2f Decision::getCentroidAtk()
{
    return centroidAtk;
}

dataState Decision::getBall()
{
    return ball;
}

void Decision::setRobots(vector<robot> r)
{
    teamRobot = r;
}

void Decision::setAreas(Point2f def, Point2f atk)
{
    centroidDef = def;
    centroidAtk = atk;
}

void Decision::setBall(dataState b)
{
    ball = b;
}

void Decision::setStrategy(int s)
{
    strategy = s;
}

void Decision::setKnnInformation(KNN k)
{
    knnfunction = k;
}

int Decision::getStrategy()
{
    return strategy;
}

void Decision::updateObjectives()
{
    if(strategy == FIXED5)
    {
        int robotFunc[5];
        teamRobot[4].setFunction(DEFENDER2);
        teamRobot[1].setFunction(DEFENDER);
        teamRobot[0].setFunction(GOALKEEPER);
        if(ball.pos.x < 125)
        {
            teamRobot[3].setFunction(FAKE9);
            teamRobot[2].setFunction(STRIKER);
        }
        else
        {
            if(flagTrocaFS == true)
            {
                if (euclidean_dist(teamRobot[2].getDataState().pos, ball.pos) < euclidean_dist(teamRobot[3].getDataState().pos, ball.pos))
                {
                    if(teamRobot[2].getDataState().pos.x < (ball.pos.x))
                    {
                        robotFunc[2] = STRIKER;
                        robotFunc[3] = FAKE9;
                    }
                    else if(teamRobot[3].getDataState().pos.x < (ball.pos.x))
                    {
                        robotFunc[2] = FAKE9;
                        robotFunc[3] = STRIKER;
                    }
                    else
                    {
                        robotFunc[lastS] = STRIKER;
                        robotFunc[lastF] = FAKE9;
                    }
                }
                else
                {
                    if(teamRobot[3].getDataState().pos.x < (ball.pos.x))
                    {
                        robotFunc[3] = STRIKER;
                        robotFunc[2] = FAKE9;
                    }
                    else if(teamRobot[2].getDataState().pos.x < (ball.pos.x))
                    {
                        robotFunc[3] = FAKE9;
                        robotFunc[2] = STRIKER;
                    }
                    else
                    {
                        robotFunc[lastS] = STRIKER;
                        robotFunc[lastF] = FAKE9;
                    }
                }

                if(robotFunc[2] == FAKE9)
                {
                    F = 2;
                    S = 3;
                }
                if(robotFunc[2] == STRIKER)
                {
                    S = 2;
                    F = 3;
                }

                if(lastF != F)
                {
                    flagTrocaFS = false;
                    clockTrocaFS = clock();
                }

                lastF = F;
                lastS = S;
                tempT = 3;

            }
            else if(flagTrocaFS == false)
            {
                timeTrocaFS = (float) (clock() - clockTrocaFS)/CLOCKS_PER_SEC;
                timeTrocaFS = timeTrocaFS * 10;

                if(timeTrocaFS > tempT)
                {
                    flagTrocaFS = true;
                }
                else
                {
                    robotFunc[lastS] = STRIKER;
                    robotFunc[lastF] = FAKE9;
                }
            }
            teamRobot[2].setFunction(robotFunc[2]);
            teamRobot[3].setFunction(robotFunc[3]);
        }
    }
    cout<< "robo 2: "<< teamRobot[2].getFunction()<<endl;
    cout<< "robo 3: "<< teamRobot[3].getFunction()<<endl;

}

File: decision.py
from robot import *

FIXED3 = 0
FULL_ATK = 1
SAFE_FULL_ATK = 2
NS = 3
FUZZY_ = 4
KNN_ = 5
IMPOSTOR_ = 6
DOWNBELOW = 7
FUZZYT = 8
OFF = 9
CRUZAMENTO = 0 # 0 = desativado

#define SIT 4380

class Decision:
    def __init__(self):
        self.pathPoints = [Point2f(), Point2f(), Point2f()]
        self.clockChange = float()
        #self.fuzzy = Fuzzy()
        self.teamRobot = [robot(), robot(), robot()]
        self.tempT = float()
        self.centroidDef = self.centroidAtk = Point2f()
        self.ball = dataState()
        self.changePermission = bool()
        self.timeChange = float()
        self.firstTime = bool()
        self.path1 = self.path2 = int()
        self.strategy = int()
        self.swapRoles = True
        self.flagCrossing = False
        self.lastB = False
        #QTimer *timer1
        self.clockInvert = self.clockTroca = float()
        self.tempoTroca = float()
        self.clockStart = self.clockTrocaGS = self.clockTrocaFS = float()
        self.temp = self.timeTrocaGS = self.timeTrocaFS = float()
        self.flagTrocaGS = True
        self.flagTrocaFS = True
        self.LineD = 48.0
        self.G = self.D = self.S = self.F = self.L = self.M = int()
        self.lastG = self.lastD = self.lastS = self.lastM = self.lastL = self.lastF = self.lastW = int()
        self.lastCompanheiroFuzzy = 1
        self.strikerIndex = 1
        self.liberoIndex = 2
        #self.knnfunction = KNN()
        self.offLine = 115.0

        #fuzzy.start()
        self.changePermission = False
        self.firstTime = True
        
    def updateObjectives(self):
        robotFunc = [int(), int(), int()]
        if self.strategy == FIXED3:
            self.teamRobot[2].setFunction(GOALKEEPER)
            self.teamRobot[1].setFunction(DEFENDER)
            self.teamRobot[0].setFunction(STRIKER)
        elif self.strategy == FULL_ATK:
            pass
        elif self.strategy == NS:
            pass
        elif self.strategy == FUZZY_:
            pass
        elif self.strategy == KNN_:
            pass
        elif self.strategy == OFF:
            pass

    def getTeamRobots(self):
        return self.teamRobot

    def getCentroidDef(self):
        return self.centroidDef
    
    def getCentroidAtk(self):
        return self.centroidAtk

    def getBall(self):
        return self.ball

    def setRobots(self, r:list):
        self.teamRobot = r

    def setAreas(self, def_:Point2f, atk_:Point2f):
        self.centroidDef = def_
        self.centroidAtk = atk_
    
    def setBall(self, b:dataState):
        self.ball = b
    
    def setStrategy(self, s:int):
        self.strategy = s
    
    '''
    def setKnnInformation(self, k:KNN):
        self.knnfunction = k
    '''

    def getStrategy(self):
        return self.strategy
    